rat/variable_table.py

```diff
@@ -14,6 +16,104 @@ class VariableType(Enum):
     ASSIGNED_PARAM = 3
 
 
+class Tracer:
+    args_set: SortedSet[Tuple]
+    arg_types: Tuple
+
+    def __init__(self, args_set=None, arg_types=None):
+        self.args_set = args_set if args_set else SortedSet()
+        self.arg_types = arg_types
+
+    def __len__(self):
+        return len(self.args_set)
+
+    def __iter__(self):
+        for args in self.args_set:
+            yield args
+
+    def __repr__(self):
+        return f"F[{str(self.arg_types)} -> ?]"
+
+    def argument_length(self):
+        if self.arg_types:
+            return len(self.arg_types)
+        else:
+            raise Exception("Argument length called before argument length known")
+
+    def __call__(self, *args):
```

Comment:
I'm wondering if `__call__` would be necessary given we aren't calling the tracers within `eval` statements. I'd prefer a method with an explicit name like `register(*args)` instead and invoke it.

Reply:
Hmm good point. This is from the codegen era but we have more control on our tracer calls now that we're not doing codegen. Will just add a note to code and save this for VariableTable refactor whenever that happens.